### PR TITLE
add: new support for SBOL3 to GenBank conversion

### PR DESCRIPTION
**Pull request by mohitdmak**
## Changes:
* Newly added `convert_sbol3_to_genbank` method of the `GenBank_SBOL3_Converter` Class supports converting from sbol3 to Genbank file formats.
* Also, using the flag `--force-new-converter` through the `sbol2genbank` and `sbol-converter` will also now invoke the new method.
* Unit test `test_sbol3togb_1` has been added to test the conversion by directly invoking the new method.
* Unit test `test_genbank_conversion_new_converter` has been added to test the conversion by invoking it from the console script, using the previously added flag `--force-new-converter.`
* 2 New Test files: `BBa_J23101_from_sbol3_direct.gb` and `iGEM_SBOL2_imports_from_sbol3_direct.gb` has been added to aid the above 2 unit tests.

### Update:
1. Throws ValueError when the sbol3 component has more than 1 sequence. (TODO: Use more specific ErrorClass later)
2. The converter used to pick up features from a sbol3 Component in random order every time the conversion was called, leading to different file outputs each time. We discussed this in our last meeting, and I have implemented a lexicographic sorting of features of each component based on the feature location's start and endpoints. This ensures the output stays the same always for a specific input sbol3 document.
3. I have made the logger throw warnings to the console whenever, a feature translation lookup does not have any row in the CSV with the role in the given SBOL3/GenBank feature. The logger warns the user of this and mentions the default ontology term it will be using.

### Clarification Needed:
- [x]  Regarding update 1.; I have made the sequence object required while defining a BioPython `SeqRecord` class as `None` when a component does not have a sequence. Is this fine?
    * Moved to https://github.com/SynBioDex/SBOL-utilities/issues/147
- [ ] Regarding update 2.; There may be multiple feature locations for every feature; in this case, how should the sorting be implemented? Currently, it just sorts features based on the start/end points assuming there is only 1 location.
    * Moved to https://github.com/SynBioDex/SBOL-utilities/issues/148
- [x] Resolvement of [this issue](https://github.com/SynBioDex/SBOL-utilities/issues/146).
- [ ] We decided to store the `VERSION` annotation for GenBank records as "1" by default for now since the extra annotation fields are not currently stored in SBOL3 files during conversion from GenBank to SBOL3. However, this does not seem to reflect in converted GenBank files (the file shows no difference in the `VERSION` field, even after specifying it during the record's creation). 
    * I think this is covered by https://github.com/SynBioDex/SBOL-utilities/issues/150. If not, a new issue should be opened specifically for the VERSION field.

## TODO:

- [ ] A point was brought up in the last discussion to remove the `write` argument altogether in both the directions of conversion and have the logic implemented separately. We would need to discuss the details of its implementation.
- [ ] Also, I learned that features of type `SubComponents` in SBOL3 components would also need to be stored in the GenBank file. Currently, the converter ignores all features other than `SequenceFeature` while converting to GenBank. This would again need some discussion to implement.
- [ ] There can also be multiple feature locations, but it turns out storing multiple feature locations for a single GenBank feature is not that straightforward. SBOL3 allows for a simple way to store locations in an array, but doing something similar for GenBank will be required too.

### CODE REVIEW

File: sbol_utilities/conversion.py
Comment (jakebeal) [inline sbol_utilities/conversion.py:424]: You don't need to turn these into keyword arguments: it's better to make minimal changes where possible.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:65]: Right name for these is "term", not "ontology" - the ontology is the full collection of organized terms.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:64]: Make an issue to consider whether this should actually be "SO:0000001" (disagreeing with libSBOLj)

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:214]: Minor style point: this is not the usual indentation for long argument lists.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:267]: Add a filter for taking only the SequenceFeature features.
Add a TODO for SubComponent features in the future.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:271]: Todo for multiple locations

**Review by jakebeal — CHANGES_REQUESTED** [review]
Just minor changes needed

**Review by tcmitchell — CHANGES_REQUESTED** [review]
Overall this looks good. As with @jakebeal comments, just a few things to change.

File: sbol_utilities/conversion.py
Comment (mohitdmak) [reply sbol_utilities/conversion.py]: Fixed

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Oh, alright, I changed them.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: I have edited this in the latest commit, but am not sure if even it is the correct way.
Maybe we should decide on a py formatter to handle these style issues?

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py:260]: Added, we are just taking note of this currently right?

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: * My bad, removed the `obj_feat.roles[0]` there.
* Actually, the `so2gb.csv` data file has rows of format `SO:xxxxxxx,<GenBank_Term>`, and the `obj_feat_role` returns the URI (i.e `https://identifiers.org/SO:xxxxxx`). So the `.index(":",5) - 2` would be the position of index from where the URI is `SO:xxxxxxx`.
* I have now added this in a comment.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Okay, fixed

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Okay, fixed

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: * Alright, I have added these constants with comments and removed the redundant line checking for `SO_FORWARD`.
* Also, @jakebeal and I discussed in last week's meeting to raise a ValueError (ValueError for now, a custom ConverterClass Error later on) when the sbol3 feature's location's orientation is none of `SO_REVERSE` or `SO_FORWARD.`
* I have added a check for this to raise a ValueError for now, is this fine?

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Done, created [here](https://github.com/SynBioDex/SBOL-utilities/issues/146).

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (jakebeal) [inline sbol_utilities/sbol3_genbank_conversion.py:65]:
```suggestion
    DEFAULT_GB_TERM = "misc_feature"
```
Per a conversation with @cjmyers today

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (mohitdmak) [inline sbol_utilities/sbol3_genbank_conversion.py:270]: @jakebeal, the TODO is present here for multiple feature locations

Comment (jakebeal) [reply sbol_utilities/sbol3_genbank_conversion.py:270]: Huh... for some reason I was seeing only the old version. Thank you!

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (tcmitchell) [reply sbol_utilities/sbol3_genbank_conversion.py]: Looks good, thanks.

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (tcmitchell) [reply sbol_utilities/sbol3_genbank_conversion.py]: Yes, this is fine. Thanks!

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (tcmitchell) [reply sbol_utilities/sbol3_genbank_conversion.py]: Thanks for adding the explanatory comment

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (tcmitchell) [reply sbol_utilities/sbol3_genbank_conversion.py]: 👍

File: sbol_utilities/sbol3_genbank_conversion.py
Comment (tcmitchell) [reply sbol_utilities/sbol3_genbank_conversion.py]: @jakebeal I see a TODO in the updated source file.

**Review by tcmitchell — APPROVED** [review]
Thanks for making the requested changes. I am approving, and we want to be sure @jakebeal also approves before merging.

We also need issues opened for the outstanding items in the PR description.

**Review by jakebeal — APPROVED** [review]

**Pull request merged**
